activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/connector/VariablesMatchHelper.java

```diff
@@ -1,15 +1,87 @@
+        return doMapping(execution, results, variableDefinitions, inMappingProvider);
+    }
+
+    public Map<String, Object> matchOut(DelegateExecution execution, Map<String, Object> results,
+                                       List<VariableDefinition> variableDefinitions) {
+        return doMapping(execution, results, variableDefinitions, outMappingProvider);
+    }
+
+    private Map<String, Object> doMapping(DelegateExecution execution, Map<String, Object> results,
+                                        List<VariableDefinition> variableDefinitions, MappingProvider mappingProvider) {
+
+        if (variableDefinitions != null && !variableDefinitions.isEmpty()) {
+            Optional<ProcessExtensionModel> optionalProcessExtensionModel = getProcessExtension(execution);
+            if (optionalProcessExtensionModel.isPresent()) {
+                ProcessExtensionModel processExtensionModel = optionalProcessExtensionModel.get();
+                VariableMapping variableMapping = processExtensionModel.getExtensions()
+                        .getVariablesMappings()
+                        .get(execution.getCurrentActivityId());
+                if (variableMapping != null) {
+                    return matchWithMapping(results, mappingProvider.getMappings(variableMapping),
+                            convertToMap(variableDefinitions),
+                            processExtensionModel.getExtensions().getProperties(),
+                            mappingProvider.getValueProvider());
+                }
+            }
+            return matchNoMapping(results, variableDefinitions);
+        }
+        return results;
+    }
+
+    private Map<String, Object> matchWithMapping(Map<String, Object> results, Map<String, String> mappings,
+                                                 Map<String, ? extends VariableDefinition> connectorVariablesMap,
+                                                 Map<String, ? extends VariableDefinition> variableDefinitionMap,
+                                                 MappedValueProvider mappedValueProvider) {
 
+        Map<String, Object> variables = new HashMap<>();
+        connectorVariablesMap.forEach((k, v) -> {
+            if (mappings.containsKey(k)) {
+                String mappedVariableId = mappings.get(v.getId());
+                VariableDefinition mappedVariableDefinition = variableDefinitionMap.get(mappedVariableId);
+                if (mappedVariableDefinition != null) {
+                    Optional<Object> outBoundVariableValueOptional = mappedValueProvider.getValue(results, v, mappedVariableDefinition);
+                    if (outBoundVariableValueOptional.isPresent()) {
+                        variables.put(mappedVariableDefinition.getName(), outBoundVariableValueOptional.get());
+                    }
+                }
+                // TODO what if the mapped variable definition wasn't found?
```

Comment:
What to do when mapping variable isn't found in connector definition? Should through exception or simple ignore? There should be a validation done before this i.e. during modelling or at startup